### PR TITLE
More lifecycle hooks

### PR DESCRIPTION
**Pull request by 9999years**
This adds support for multiple test commands (`--test-ghci`) and before/after reload/restart hooks: `--before-reload-ghci`, `--after-reload-ghci`, `--before-restart-ghci`, and `--after-restart-ghci`. We may also consider before/after hooks for eval commands.

### CODE REVIEW

**Comment by 9999years** [comment]
Current dependencies on/for this PR:
* main
  * **PR #73** <a href="https://app.graphite.dev/github/pr/MercuryTechnologies/ghcid-ng/73" target="_blank">Graphite</a> 
    * **PR #74** <a href="https://app.graphite.dev/github/pr/MercuryTechnologies/ghcid-ng/74" target="_blank">Graphite</a> 
      * **PR #75** <a href="https://app.graphite.dev/github/pr/MercuryTechnologies/ghcid-ng/75" target="_blank">Graphite</a> 
        * **PR #76** <a href="https://app.graphite.dev/github/pr/MercuryTechnologies/ghcid-ng/76" target="_blank">Graphite</a> 
          * **PR #77** <a href="https://app.graphite.dev/github/pr/MercuryTechnologies/ghcid-ng/77" target="_blank">Graphite</a> 
            * **PR #78** <a href="https://app.graphite.dev/github/pr/MercuryTechnologies/ghcid-ng/78" target="_blank">Graphite</a> 
              * **PR #86** <a href="https://app.graphite.dev/github/pr/MercuryTechnologies/ghcid-ng/86" target="_blank">Graphite</a> 
                * **PR #92** <a href="https://app.graphite.dev/github/pr/MercuryTechnologies/ghcid-ng/92" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/MercuryTechnologies/ghcid-ng/92?utm_source=stack-comment).

**Comment by 9999years** [comment]
Current dependencies on/for this PR:
* main
  * **PR #73** <a href="https://app.graphite.dev/github/pr/MercuryTechnologies/ghcid-ng/73" target="_blank">Graphite</a> 
    * **PR #74** <a href="https://app.graphite.dev/github/pr/MercuryTechnologies/ghcid-ng/74" target="_blank">Graphite</a> 
      * **PR #75** <a href="https://app.graphite.dev/github/pr/MercuryTechnologies/ghcid-ng/75" target="_blank">Graphite</a> 
        * **PR #76** <a href="https://app.graphite.dev/github/pr/MercuryTechnologies/ghcid-ng/76" target="_blank">Graphite</a> 
          * **PR #77** <a href="https://app.graphite.dev/github/pr/MercuryTechnologies/ghcid-ng/77" target="_blank">Graphite</a> 
            * **PR #78** <a href="https://app.graphite.dev/github/pr/MercuryTechnologies/ghcid-ng/78" target="_blank">Graphite</a> 
              * **PR #86** <a href="https://app.graphite.dev/github/pr/MercuryTechnologies/ghcid-ng/86" target="_blank">Graphite</a> 
                * **PR #92** <a href="https://app.graphite.dev/github/pr/MercuryTechnologies/ghcid-ng/92" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/MercuryTechnologies/ghcid-ng/92?utm_source=stack-comment).

File: src/cli.rs [hunk 1]
```diff
@@ -132,6 +120,55 @@ pub struct LoggingOpts {
...
 }
 
+/// Lifecycle hooks.
+///
+/// These are commands (mostly `ghci` commands) to run at various points in the `ghcid-ng`
+/// lifecycle.
+#[derive(Debug, Clone, clap::Args)]
+#[clap(next_help_heading = "Lifecycle hooks")]
+pub struct HookOpts {
+    /// `ghci` commands which runs tests, like `TestMain.testMain`. If given, these commands will be
```
Comment (Gabriella439) [inline src/cli.rs:130]:
```suggestion
    /// `ghci` commands which run tests, like `TestMain.testMain`. If given, these commands will be
```

File: src/ghci/mod.rs [hunk 2]
```diff
@@ -370,6 +378,11 @@ impl Ghci {
                 .await?;
         }
 
+        for command in &self.opts.hooks.after_reload_ghci {
```
Comment (Gabriella439) [inline src/ghci/mod.rs:381]: should this be slightly further up (inside the `if !actions.needs_reload.is_empty() {` expression)?

File: src/ghci/mod.rs [hunk 3]
```diff
@@ -330,16 +325,29 @@ impl Ghci {
     /// This may fully restart the `ghci` process.
     #[instrument(skip_all, level = "debug")]
     pub async fn reload(&mut self, events: Vec<FileEvent>) -> miette::Result<()> {
+        for command in &self.opts.hooks.before_reload_ghci {
```
Comment (Gabriella439) [inline src/ghci/mod.rs:328]: should this be further down (inside the `if !actions.needs_reload.is_empty() {` expression)?

**Comment by Gabriella439** [comment]
wait, the last commit wasn't the change i was expecting.  i was thinking of something more like modifying the existing `if !actions.needs_reload.is_empty() {` expression to be like this:

```rust
        if !actions.needs_reload.is_empty() {
            // pre-reload command goes here

            tracing::info!(
                "Reloading ghci due to changed modules:\n{}",
                format_bulleted_list(&actions.needs_reload)
            );
            let messages = self.stdin.reload(&mut self.stdout).await?;

            // post-reload command goes here

            if let Some(CompilationResult::Err) = self.process_ghc_messages(messages).await? {
                compilation_failed = true;
            }
            self.refresh_eval_commands_for_paths(&actions.needs_reload)
                .await?;
        }
```

**Review by Gabriella439 — APPROVED** [review]

**Pull request merged**